codec/src/test/java/io/netty/handler/codec/compression/Lz4FrameEncoderTest.java

```diff
@@ -52,4 +82,128 @@ protected ByteBuf decompress(ByteBuf compressed, int originalLength) throws Exce
+        Checksum checksum = XXHashFactory.fastestInstance().newStreamingHash32(DEFAULT_SEED).asChecksum();
+        Lz4FrameEncoder encoder = new Lz4FrameEncoder(LZ4Factory.fastestInstance(), true,
+                                                      blockSize,
+                                                      checksum,
+                                                      maxEncodeSize);
+        encoder.handlerAdded(ctx);
+        return encoder;
+    }
+
+    @Test (expected = EncoderException.class)
+    public void testAllocateDirectBuffer_OverflowsOutputSize() {
+        final int maxEncodeSize = Integer.MAX_VALUE;
+        Lz4FrameEncoder encoder = newEncoder(Lz4Constants.DEFAULT_BLOCK_SIZE, maxEncodeSize);
+        ByteBuf buf = null;
+        try
+        {
+            buf = ByteBufAllocator.DEFAULT.buffer(maxEncodeSize, maxEncodeSize);
+            buf.writerIndex(maxEncodeSize);
+            encoder.allocateBuffer(ctx, buf, false);
+        } catch (OutOfDirectMemoryError e) {
+            // this sometimes fails on the continuous integration server if we try to allocate Integer.MAX_VALUE.
+            // ignore the failure in that case
+        } finally {
+            if (buf != null) {
+                buf.release();
+            }
+        }
+    }
+
+    @Test
+    public void testFlush() {
+        Lz4FrameEncoder encoder = new Lz4FrameEncoder();
+        EmbeddedChannel channel = new EmbeddedChannel(encoder);
+        int size = 27;
+        ByteBuf buf = ByteBufAllocator.DEFAULT.buffer(size, size);
+        buf.writerIndex(size);
+        Assert.assertEquals(0, encoder.getBuffer().readableBytes());
+        channel.write(buf);
+        Assert.assertTrue(channel.outboundMessages().isEmpty());
+        Assert.assertEquals(size, encoder.getBuffer().readableBytes());
+        channel.flush();
+        Assert.assertTrue(channel.finish());
+        Assert.assertTrue(channel.releaseOutbound());
+        Assert.assertFalse(channel.releaseInbound());
+    }
+
+    @Test
+    public void testAllocatingAroundBlockSize() {
+        int blockSize = 100;
+        Lz4FrameEncoder encoder = newEncoder(blockSize, Lz4FrameEncoder.DEFAULT_MAX_ENCODE_SIZE);
+        EmbeddedChannel channel = new EmbeddedChannel(encoder);
+
+        int size = blockSize - 1;
+        ByteBuf buf = ByteBufAllocator.DEFAULT.buffer(size, size);
+        buf.writerIndex(size);
+        Assert.assertEquals(0, encoder.getBuffer().readableBytes());
+        channel.write(buf);
+        Assert.assertEquals(size, encoder.getBuffer().readableBytes());
+
+        int nextSize = size - 1;
+        buf = ByteBufAllocator.DEFAULT.buffer(nextSize, nextSize);
+        buf.writerIndex(nextSize);
+        channel.write(buf);
+        Assert.assertEquals(size + nextSize - blockSize, encoder.getBuffer().readableBytes());
+
+        channel.flush();
+        Assert.assertEquals(0, encoder.getBuffer().readableBytes());
+    }
```

Comment:
Make sure you call `channel.finishAndReleaseAll();` or verify the channel has some output buffers like you do in the previous `testFlush` test.